app/jinja_filters.py

```diff
@@ -306,25 +306,23 @@ def _get_first_non_empty_item(items):
     :param items: anything that is iterable
     :return: first non empty value
 
-    In this filter the following values are considered false:
+    In this filter the following values are considered non empty:
     - None
-    - False
     - any empty sequence, for example, '', (), [].
     - any empty mapping, for example, {}.
-    - instances of user-defined classes, if the class defines a __bool__() method, when that method returns bool value False.
 
      Note: to guarantee the returned element is actually the first non empty element in the iterable,
     'items' must be a data structure that preserves order, ie tuple, list etc.
     If order is not important, this can be reused to return `one of` the elements which is non empty.
+
+    This filter will treat zero of any numeric type for example, 0, 0.0, 0j and boolean 'False'
+    as a valid item since they are naturally 'falsy' in Python but not empty.
+
+    Note: Booleans are a subtype of integers. Zero of any numeric type 'is not False' but 'equals False'.
+    Reference: https://docs.python.org/release/3.4.2/library/stdtypes.html?highlight=boolean#boolean-values
     """
     for item in items:
-        if item:
-            return item
-
-        # The below statement will treat zero of any numeric type for example, 0, 0.0, 0j as a valid item since they are naturally 'falsy' in Python.
-        # Note: Booleans are a subtype of integers. Zero of any numeric type 'is not False' but 'equals False'.
-        # Reference: https://docs.python.org/release/3.4.2/library/stdtypes.html?highlight=boolean#boolean-values
-        if item is not False and item == 0:
+        if item or item is False or item == 0:
             return item
 
     raise Exception('No valid items provided.')
```

Comment:
Discussed a bit in person, but I think this should still raise an exception. The behaviour of the filter isn't really defined for that situation, and I don't think it can be in a reasonable way, since the use cases are varied (Do you want an empty string / "Unknown" / None etc).

I would suggest passing a list of things with no 'valid' values is an improper use of the filter, indicating a schema error. I don't think we can reasonably expect to catch an error like this in validator, and would therefore expect an exception to be raised.